### PR TITLE
VSD: Simplifying abstract_objectt::write, collapsing inheritance hierarchy

### PR DESCRIPTION
**Pull request by jezhiggins**
This PR simplifies the `write` member function implementation in `abstract_objectt` subclasses. It collapses the inheritance hierarchy for classes modelling aggregate values (arrays, structs, unions), eliminating duplication. 

Several classes were named to better reflect their purpose - eg `union_abstract_objectt` became `two_value_union_abstract_objectt`, and some new classes introduced to in place of a previously existing base class, eg `two_value_array_abstract_objectt` in place of the eliminated `array_abtract_objectt`. 

This allowed further work to combine `interval_array_abstract_objectt` and `constant_array_abstract_objectt` together. The combined implementation has been renamed `full_array_abstract_objectt`, and the implementation refactored for reading comprehension. 

This PR is behaviour preserving. No need functionality is introduced.

### CODE REVIEW

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (martin-cs) [inline src/analyses/variable-sensitivity/abstract_enviroment.cpp:62]: It is true that it is handled the same but it is for quite different reasons.  This is not necessarily a problem, but it is something to be aware of.

File: src/analyses/variable-sensitivity/abstract_aggregate_object.h
Comment (martin-cs) [inline src/analyses/variable-sensitivity/abstract_aggregate_object.h]: This is probably a bit more template-y than we normally write but under the circumstances I can see this reduces an amount of duplication.

File: src/analyses/variable-sensitivity/interval_array_abstract_object.cpp
Comment (martin-cs) [inline src/analyses/variable-sensitivity/interval_array_abstract_object.cpp]: Although not touched by this PR, this case really should result in the interval being top.

File: src/analyses/variable-sensitivity/abstract_value_object.h
Comment (martin-cs) [inline src/analyses/variable-sensitivity/abstract_value_object.h:72]: New line?

File: src/analyses/variable-sensitivity/abstract_value_object.h
Comment (martin-cs) [inline src/analyses/variable-sensitivity/abstract_value_object.h]: I'm not sure I follow why this is needed...

File: src/analyses/variable-sensitivity/interval_abstract_value.cpp
Comment (martin-cs) [inline src/analyses/variable-sensitivity/interval_abstract_value.cpp]: I get that if the interval is bottom you should get an empty range but I'm not sure I get why this is the case for top or if the interval is open at one end.

File: unit/analyses/variable-sensitivity/abstract_object/index_range.cpp
Comment (martin-cs) [inline unit/analyses/variable-sensitivity/abstract_object/index_range.cpp:46]: If it tis top then it should be everything so this is definitely an improvement but I'm not quite sure what the nil is supposed to describe.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (martin-cs) [inline src/analyses/variable-sensitivity/abstract_object.h:446]: Good tidy up.

File: src/analyses/variable-sensitivity/full_array_abstract_object.cpp
Comment (martin-cs) [inline src/analyses/variable-sensitivity/full_array_abstract_object.cpp:251]: It's not entirely clear to me why this is needed.  Can't you just delegate the write to the relevant `abstract_object` and then update the relevant entry with whatever it returns?

File: src/analyses/variable-sensitivity/full_array_abstract_object.cpp
Comment (martin-cs) [inline src/analyses/variable-sensitivity/full_array_abstract_object.cpp:354]: It may not change the value that is stored but breaking sharing is a serious concern.  Please check that the sharing map does actually handle that case correctly.

**Review by martin-cs — APPROVED** [review]
I don't see any immediate show stoppers but this is quite a substantial refactor and it would be good to work through the code-base with you before we merge this PR.

File: src/analyses/variable-sensitivity/abstract_aggregate_object.h
Comment (jezhiggins) [reply src/analyses/variable-sensitivity/abstract_aggregate_object.h]: It's compile type polymorphism rather than run-time. We don't need, actually don't want, run-time polymorphism here.

File: src/analyses/variable-sensitivity/abstract_value_object.h
Comment (jezhiggins) [reply src/analyses/variable-sensitivity/abstract_value_object.h]: I'd initially expected `abstract_value_objectt` to be a template in a similar way to `abstract_aggregate_objectt`. Didn't turn out that way, although may yet with the next bit of work. Happy to remove this now if you'd prefer.

File: src/analyses/variable-sensitivity/interval_abstract_value.cpp
Comment (jezhiggins) [reply src/analyses/variable-sensitivity/interval_abstract_value.cpp]: It doesn't make sense to me either, but this is the existing behaviour.

File: unit/analyses/variable-sensitivity/abstract_object/index_range.cpp
Comment (jezhiggins) [reply unit/analyses/variable-sensitivity/abstract_object/index_range.cpp:46]: The index_range is only used to during array access, and this supports the existing behaviour. Returning a nil expression here takes us down to https://github.com/jezhiggins/cbmc/blob/b618d2e3703434aa0cfb4d1c991c867b1dea9955/src/analyses/variable-sensitivity/full_array_abstract_object.cpp#L298 
It is a special case, but it's isolated here rather than having to be explicitly handed in potentially several places elsewhere. As we look to expand other value representations, the more we can keep within those representations, the better.

File: src/analyses/variable-sensitivity/full_array_abstract_object.cpp
Comment (jezhiggins) [reply src/analyses/variable-sensitivity/full_array_abstract_object.cpp:251]: No, because once we get to the end of the stack we need to worry about whether it's a merging write or not. Higher up the stack, we don't need to do that. 

We have a similar decision point in `full_struct_abstract_objectt`, and I assume we'll have one in our union representation when we fill that out. When we get there, it's entirely possible we'll be able to collapse the implementations.

**Comment by martin-cs** [comment]
I've had a discussion with @jezhiggins and I am in favour of merging this because:

1. It is at least an intermediate step towards sorting this code out.
2. It is code that I requested we merge "to be improved in tree".
3. It makes things simpler.
4. It is a pure refactoring.  Jez has very intentionally preserved functionality, even when questionable.

**Comment by martin-cs** [comment]
@tautschnig @peterschrammel This is part of the refactor and tidy up of VSD.  I don't expect you to look at that part of the commit, i think it is a net improvement and there is more to do.  Please could you have a look at the changes in other parts of the tree, notably `util/' which are blocking the merge.

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (tautschnig) [inline src/analyses/variable-sensitivity/abstract_enviroment.cpp]: Seems surprising?

File: src/analyses/variable-sensitivity/abstract_environment.cpp
Comment (tautschnig) [inline src/analyses/variable-sensitivity/abstract_environment.cpp]: Why is this file being included?

File: unit/analyses/variable-sensitivity/constant_abstract_value/constant_index_range.cpp
Comment (tautschnig) [inline unit/analyses/variable-sensitivity/constant_abstract_value/constant_index_range.cpp]: Quoting `CODING_GUIDELINES.md`: "Each source and header file must start with a comment block stating the author."

File: src/analyses/variable-sensitivity/abstract_value_object.cpp
Comment (tautschnig) [inline src/analyses/variable-sensitivity/abstract_value_object.cpp:9]: See earlier comment: please make sure all files include a header stating the author.

File: src/analyses/variable-sensitivity/full_array_abstract_object.cpp
Comment (tautschnig) [inline src/analyses/variable-sensitivity/full_array_abstract_object.cpp:17]: Should likely be `static` as only required within this file.

File: src/util/sharing_map.h
Comment (tautschnig) [inline src/util/sharing_map.h:467]: Would you mind adding documentation? It may be ok to put the three methods in a group, but there should be some sort of documentation.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (tautschnig) [inline src/analyses/variable-sensitivity/abstract_object.h]: How about changing the interface to make `merge` return a pair?

**Review by tautschnig — APPROVED** [review]
Some administrative comments only, I have browsed (but not properly reviewed) the code in analyses/ and focussed on the single change in src/util/ only.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (jezhiggins) [reply src/analyses/variable-sensitivity/abstract_object.h]: I did try it, but it didn't really make the code any clearer. If structured bindings were available, I'd absolutely have gone for it.

File: src/analyses/variable-sensitivity/abstract_environment.cpp
Comment (jezhiggins) [reply src/analyses/variable-sensitivity/abstract_environment.cpp]: Good question - looking at now, I probably included for some debug/dev code which didn't need to be committed. I'll tidy this up.

**Pull request merged**
